sdk/core/azure-core/inc/nullable.hpp

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <cstdlib> // for abort
+#include <new> // for placement new
+#include <type_traits>
+#include <utility> // for swap and move
+
+namespace Azure { namespace Core {
+  namespace Details {
+    struct NontrivialDummyType
+    {
+      constexpr NontrivialDummyType() noexcept {}
+    };
+  }
+
+  template <class T> class Nullable {
+    union
+    {
+      Details::NontrivialDummyType m_disengaged; // due to constexpr rules for the default constructor
+      T m_value;
+    };
+
+    bool m_hasValue;
+
+    void Destroy() noexcept /* enforces termination */
+    {
+      if (m_hasValue)
+      {
+        m_value.~T();
+        m_hasValue = false;
+      }
+    }
+
+  public:
+    constexpr Nullable() : m_disengaged{}, m_hasValue(false) {}
+    constexpr Nullable(const T& initialValue) : m_value(initialValue), m_hasValue(true) {}
+
+    Nullable(const Nullable& other) noexcept(std::is_nothrow_copy_constructible<T>::value)
+        : m_hasValue(other.m_hasValue)
+    {
+      if (m_hasValue)
+      {
+        ::new (static_cast<void*>(&m_value)) T(other.m_value);
+      }
+    }
+
+    Nullable(Nullable&& other) noexcept(std::is_nothrow_move_constructible<T>::value)
+        : m_hasValue(other.m_hasValue)
+    {
+      if (m_hasValue)
+      {
+        ::new (static_cast<void*>(&m_value)) T(std::move(other.m_value));
+      }
+    }
+
+    ~Nullable()
+    {
+      if (m_hasValue)
+      {
+        m_value.~T();
+      }
+    }
+
+    void Reset() noexcept /* enforces termination */ { Destroy(); }
```

Comment:
Technically the termination enforcement is happening in Destroy... OTOH this is the public one. Up to you.